Adds Lucas-Kanade optical flow algorithm

arijitkar98 opened the pull request:
@zygmuntszpak  For the initial implementation I have used a `Coordinate` struct type. I am not sure if this is the correct way to go. 

Also, I need some help regarding which versions of packages to put in `REQUIRE` and how to set the build status in `README`.

I will tests and comments to the code tomorrow.

arijitkar98 committed:
Updates README.md

arijitkar98 committed:
Exports Lucas Kanade Algorithm

arijitkar98 committed:
Modifies function optical_flow

arijitkar98 committed:
Adds Coordinate type and related functions in core.jl

arijitkar98 committed:
Adds lucas kanade algo type

arijitkar98 committed:
Adds Lucas Kanade optical flow algorithm implementation

arijitkar98 committed:
Makes changes to API

arijitkar98 committed:
Adds comments to function

arijitkar98 committed:
Adds basic translation tests

arijitkar98 committed:
Updates REQUIRE

zygmuntszpak commented on test/optical_flow.jl:
```diff
@@ -0,0 +1,115 @@
+
+@testset "lucas-kanade" begin
+
+    #Basic translations (Horizontal)
```
You can use the `warp` function to create a translated version of `img1` instead of doing this manually. 
For example,
```julia
using TestImages,CoordinateTransformations, Rotations, StaticArrays
img = Gray.(testimage("mandrill"))
trans = Translation(1.0, 0.0)
imgtran= warp(img, trans)
indices(imgtran)
```
See [the documentation on indexing](https://juliaimages.github.io/latest/indexing.html) for another example of the `warp` function.

arijitkar98 replied:
I have modified two of these manual initialisations to `warp` and kept one as it is so as to test the function for normal arrays (not `OffsetArray`) as well

zygmuntszpak commented on test/optical_flow.jl:
```diff
@@ -0,0 +1,115 @@
+    @test ((lost_points/size(a)[1])*100) < max_lost_points_percentage
+
+
+    # Basic translations (Vertical)
```
Use `warp` here as well.

arijitkar98 replied:
Done

zygmuntszpak commented on test/optical_flow.jl:
```diff
@@ -0,0 +1,115 @@
+    @test ((lost_points/size(a)[1])*100) < max_lost_points_percentage
+
+
+    # Basic translations (Both)
```
`warp` here too.

arijitkar98 replied:
Done

zygmuntszpak commented on test/optical_flow.jl:
```diff
@@ -0,0 +1,115 @@
+    end
+    test_img1 = Gray.(img1)
+    test_img2 = Gray.(img2)
+    corners = imcorner(test_img1, method=shi_tomasi)
```
Instead of introducing the `Coordinate` type, perhaps just use `SVector` directly.  So instead of storing an array of `Coordinate` types, you will store an array of `SVector` instead. An `SVector` is allocated on the stack, and for the types of operations that you will perform I suspect that it will lead to much faster code.

The code for creating  the array of coordinates can be written much more concisely as follows:
```julia
r, c, _ = findnz(corners)
corner_coords = map((ri, ci) -> SVector{2}(ri, ci), r, c)
pts = rand(corner_coords, (number_test_pts,))
```

I discovered this neat approach on[ line 58 of core.jl from ImageFeatures](https://github.com/JuliaImages/ImageFeatures.jl/blob/master/src/core.jl).

arijitkar98 replied:
Done

zygmuntszpak commented on src/lucas_kanade.jl:
```diff
@@ -0,0 +1,234 @@
+    status_array = trues(size(algo.prev_points))
+    error_array = zeros(Float64, size(algo.prev_points))
+
+    for i = 1:size(algo.prev_points)[1]
```
Since your data structure is an array of coordinates, you can just do
```julia
for i = 1:length(algo.prev_points)
```
instead of 
```julia
for i = 1:size(algo.prev_points)[1]
```
for your loops.

arijitkar98 replied:
Done

zygmuntszpak commented on src/lucas_kanade.jl:
```diff
@@ -0,0 +1,234 @@
+    second_pyr = gaussian_pyramid(second_img, algo.max_level, 2, 1.0)
+
+    # Initilisation of the output arrays
+    guess_flow = Array{Coordinate{Float64},1}(size(algo.prev_points)[1])
```
I think you can change to this to 
```julia
guess_flow = similar(algo.prev_points)
final_flow = similar(algo.prev_points)
```

arijitkar98 replied:
I need `guess_flow` and `final_flow` to be `Array{MVector{2, Float64}, 1}` and thus have initialised them manually.

zygmuntszpak commented:
At the the pull-request fails with errors for me
```julia 
              _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.2 (2017-12-13 18:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

julia> Pkg.test("ImageTracking")
INFO: Computing test dependencies for ImageTracking...
INFO: No packages to install, update or remove
INFO: Testing ImageTracking
lucas-kanade: Error During Test
  Got an exception of type MethodError outside of a @test
  MethodError: no method matching getindex(::ColorTypes.Gray{Float64})
  Stacktrace:
   [1] macro expansion at /home/zygmunt/.julia/v0.6/StaticArrays/src/broadcast.jl:173 [inlined]
   [2] _broadcast at /home/zygmunt/.julia/v0.6/StaticArrays/src/broadcast.jl:134 [inlined]
   [3] broadcast_c at /home/zygmunt/.julia/v0.6/StaticArrays/src/broadcast.jl:37 [inlined]
   [4] broadcast at ./broadcast.jl:455 [inlined]
   [5] + at /home/zygmunt/.julia/v0.6/ColorVectorSpace/src/ColorVectorSpace.jl:322 [inlined]
   [6] _rightborder!(::Array{ColorTypes.Gray{Float64},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}, ::CartesianIndex{0}, ::UnitRange{Int64}, ::CartesianIndex{1}, ::ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}}) at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:1393
   [7] _imfilter_dim!(::ComputationalResources.CPU1{ImageFiltering.Algorithm.IIR}, ::Array{ColorTypes.Gray{Float64},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}, ::CartesianRange{CartesianIndex{0}}, ::Base.OneTo{Int64}, ::CartesianRange{CartesianIndex{1}}, ::ImageFiltering.Fill{ColorTypes.Gray{Float64},0}) at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:1329
   [8] _imfilter_inplace_tuple! at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:1272 [inlined]
   [9] imfilter!(::ComputationalResources.CPU1{ImageFiltering.Algorithm.IIR}, ::Array{ColorTypes.Gray{Float64},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Tuple{ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9},ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}}, ::ImageFiltering.Fill{ColorTypes.Gray{Float64},0}) at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:1229
   [10] imfilter_na_separable! at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:1452 [inlined]
   [11] _imfilter_na! at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:687 [inlined]
   [12] imfilter!(::ComputationalResources.CPU1{ImageFiltering.Algorithm.IIR}, ::Array{ColorTypes.Gray{Float64},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Tuple{ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9},ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}}, ::ImageFiltering.NA{0}) at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:662
   [13] imfilter!(::Array{ColorTypes.Gray{Float64},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Tuple{ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9},ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}}, ::ImageFiltering.NA{0}, ::ImageFiltering.Algorithm.IIR) at /home/zygmunt/.julia/v0.6/ImageFiltering/src/imfilter.jl:612
   [14] gaussian_pyramid(::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Int64, ::Int64, ::Tuple{ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9},ImageFiltering.KernelFactors.TriggsSdika{Float64,3,3,9}}) at /home/zygmunt/.julia/v0.6/Images/src/algorithms.jl:802
   [15] optflow(::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::ImageTracking.LK{Int64,Float64,Bool}) at /home/zygmunt/.julia/v0.6/ImageTracking/src/lucas_kanade.jl:53
   [16] optical_flow(::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2}, ::ImageTracking.LK{Int64,Float64,Bool}) at /home/zygmunt/.julia/v0.6/ImageTracking/src/optical_flow.jl:21
   [17] macro expansion at /home/zygmunt/.julia/v0.6/ImageTracking/test/optical_flow.jl:58 [inlined]
   [18] macro expansion at ./test.jl:860 [inlined]
   [19] anonymous at ./<missing>:?
   [20] include_from_node1(::String) at ./loading.jl:576
   [21] include(::String) at ./sysimg.jl:14
   [22] include_from_node1(::String) at ./loading.jl:576
   [23] include(::String) at ./sysimg.jl:14
   [24] process_options(::Base.JLOptions) at ./client.jl:305
   [25] _start() at ./client.jl:371

```

However, by looking at Travis it seems that it did actually execute successfully on `0.6` so I'm somewhat confused. Is it building fine for you?

arijitkar98 commented:
Yes, it is building at my end. I will incorporate these changes by tonight

arijitkar98 committed:
Modifies lucas-kanade function and corresponding tests

(1) Updated most of the code SVectors and MVectors
(2) Corrected percentages in tests
(3) Corrected tracking lost condition

arijitkar98 committed:
Fixed random seed in test file for lucas-kanade

arijitkar98 committed:
Adds OffsetArrays capability to the function

arijitkar98 commented:
@zygmuntszpak I have added the `OffsetArrays` capability. This build will fail until the [PR #722](https://github.com/JuliaImages/Images.jl/pull/722) in Images.jl gets merged.

arijitkar98 committed:
Updates tests
